File: python3/bron_kerbosch2.py
# coding: utf-8

from bron_kerbosch_pivot import pick_arbitrary, visit
from graph import UndirectedGraph
from reporter import Reporter


def bron_kerbosch2(graph: UndirectedGraph, reporter: Reporter) -> None:
    """Bron-Kerbosch algorithm with arbitrarily chosen pivot"""
    if candidates := graph.connected_vertices():
        visit(graph=graph,
              reporter=reporter,
              initial_pivot_choice=pick_arbitrary,
              further_pivot_choice=pick_arbitrary,
              candidates=candidates,
              excluded=set(),
              clique=[])

File: scala/src/main/scala/bron_kerbosch2.scala
import bron_kerbosch_pivot.PivotChoice.Arbitrary
import bron_kerbosch_pivot.visit

import scala.collection.{immutable, mutable}

class bron_kerbosch2[Vertex] extends bron_kerbosch_algorithm[Vertex] {
  def explore(
      graph: UndirectedGraph[Vertex],
      reporter: immutable.Iterable[Vertex] => Unit
  ): Unit = {
    val candidates = graph.connected_vertices().to(mutable.Set)
    if (candidates.nonEmpty) {
      visit(
        graph,
        reporter,
        Arbitrary,
        Arbitrary,
        candidates,
        mutable.Set(),
        immutable.List()
      )
    } else {
      immutable.Iterable()
    }
  }
}
